File: backend/app/services/project_service.py
from sqlalchemy.orm import Session

from app.models.project import Project
from app.schemas.project import (
    ProjectCreate,
    ProjectUpdate,
)


def create_project(
    db: Session,
    project_data: ProjectCreate,
    user_id: int,
):
    project = Project(
        name=project_data.name,
        description=project_data.description,
        user_id=user_id,
    )

    db.add(project)
    db.commit()
    db.refresh(project)

    return project


def get_projects_by_user(
    db: Session,
    user_id: int,
):
    return (
        db.query(Project)
        .filter(
            Project.user_id == user_id
        )
        .order_by(
            Project.created_at.desc()
        )
        .all()
    )


def get_project_by_id(
    db: Session,
    project_id: int,
    user_id: int,
):
    return (
        db.query(Project)
        .filter(
            Project.id == project_id,
            Project.user_id == user_id,
        )
        .first()
    )


def update_project(
    db: Session,
    project: Project,
    project_data: ProjectUpdate,
):
    update_data = project_data.model_dump(
        exclude_unset=True
    )

    for field, value in update_data.items():
        setattr(
            project,
            field,
            value,
        )

    db.commit()
    db.refresh(project)

    return project


def delete_project(
    db: Session,
    project: Project,
):
    db.delete(project)
    db.commit()

File: frontend/src/services/projectService.ts
import api from "../api/axios";

export interface ProjectCreate {
  name: string;
  description?: string;
}

export interface ProjectUpdate {
  name?: string;
  description?: string;
}

export const getProjects = async () => {
  const response = await api.get("/projects");
  return response.data;
};

export const createProject = async (
  data: ProjectCreate
) => {
  const response = await api.post(
    "/projects",
    data
  );

  return response.data;
};

export const getProject = async (
  projectId: number
) => {
  const response = await api.get(
    `/projects/${projectId}`
  );

  return response.data;
};

export const updateProject = async (
  projectId: number,
  data: ProjectUpdate
) => {
  const response = await api.patch(
    `/projects/${projectId}`,
    data
  );

  return response.data;
};

export const deleteProject = async (
  projectId: number
) => {
  await api.delete(
    `/projects/${projectId}`
  );
};